src/crystal/system/unix/getrandom.cr

```diff
@@ -1,116 +1,48 @@
-{% skip_file unless flag?(:linux) %}
-
-require "c/unistd"
-require "./syscall"
-
-{% if flag?(:interpreted) %}
-  lib LibC
-    fun getrandom(buf : Void*, buflen : SizeT, flags : UInt32) : LibC::SSizeT
-  end
-
-  module Crystal::System::Syscall
-    GRND_NONBLOCK = 1u32
-
-    # TODO: Implement syscall for interpreter
-    def self.getrandom(buf : UInt8*, buflen : LibC::SizeT, flags : UInt32) : LibC::SSizeT
-      LibC.getrandom(buf, buflen, flags)
+require "c/sys/random"
+
+{% if !LibC.has_method?(:getrandom) %}
+  # getrandom isn't available in android before API LEVEL 28
+  require "./urandom"
+{% else %}
+  module Crystal::System::Random
+    # Reads n random bytes using the Linux `getrandom(2)` syscall.
+    def self.random_bytes(buffer : Bytes) : Nil
+      getrandom(buffer)
     end
-  end
-{% end %}
-
-module Crystal::System::Random
-  @@initialized = false
-  @@getrandom_available = false
-  @@urandom : ::File?
-
-  private def self.init
-    @@initialized = true
-
-    if has_sys_getrandom
-      @@getrandom_available = true
-    else
-      urandom = ::File.open("/dev/urandom", "r")
-      return unless urandom.info.type.character_device?
-
-      urandom.close_on_exec = true
-      urandom.read_buffering = false # don't buffer bytes
-      @@urandom = urandom
-    end
-  end
 
-  private def self.has_sys_getrandom
-    sys_getrandom(Bytes.new(16))
-    true
-  rescue
-    false
-  end
-
-  # Reads n random bytes using the Linux `getrandom(2)` syscall.
-  def self.random_bytes(buf : Bytes) : Nil
-    init unless @@initialized
-
-    if @@getrandom_available
-      getrandom(buf)
-    elsif urandom = @@urandom
-      urandom.read_fully(buf)
-    else
-      raise "Failed to access secure source to generate random bytes!"
+    def self.next_u : UInt8
+      buffer = uninitialized UInt8
+      getrandom(pointerof(buffer).to_slice(1))
+      buffer
     end
-  end
 
-  def self.next_u : UInt8
-    init unless @@initialized
-
-    if @@getrandom_available
-      buf = uninitialized UInt8
-      getrandom(pointerof(buf).to_slice(1))
-      buf
-    elsif urandom = @@urandom
-      urandom.read_byte.not_nil!
-    else
-      raise "Failed to access secure source to generate random bytes!"
-    end
-  end
+    # Reads n random bytes using the Linux `getrandom(2)` syscall.
+    private def self.getrandom(buffer)
+      # getrandom(2) may only read up to 256 bytes at once without being
+      # interrupted or returning early
+      chunk_size = 256
 
-  # Reads n random bytes using the Linux `getrandom(2)` syscall.
-  private def self.getrandom(buf)
-    # getrandom(2) may only read up to 256 bytes at once without being
-    # interrupted or returning early
-    chunk_size = 256
+      while buffer.size > 0
+        if buffer.size < chunk_size
+          chunk_size = buffer.size
+        end
```

Comment:
**issue:** It seems this doesn't have any effect?
Maybe we should use `buffer.size.clamp(..256)` for the length argument?